src/main/java/com/google/summit/serialization/Serializer.kt

```diff
@@ -0,0 +1,135 @@
+ * parameters.
+ *
+ * Deserializing JSON without source locations is unsupported and may result in undefined behavior.
+ *
+ * Usage:
+ * ```
+ * val cu: CompilationUnit = ...
+ * val serializer = Serializer()
+ * val serialized: String = serializer.serialize(cu)
+ * val deserialized: CompilationUnit? = serializer.deserialize(CompilationUnit::class.java, serialized)
+ * ```
+ *
+ * @param format whether to add indentation and formatting to the output
+ * @param removeLocations whether to filter [source locations][SourceLocation] from serialization
+ */
+class Serializer(format: Boolean = false, removeLocations: Boolean = false) {
+  private companion object {
+    /** Recursively finds the subtypes of [baseType] that are known at compile time. */
+    fun <T : Any> findSubtypes(baseType: KClass<T>): List<KClass<out T>> {
+
+      fun find(cl: KClass<out T>): List<KClass<out T>> =
+        listOf(cl) + cl.sealedSubclasses.flatMap { find(it) }
+
+      return find(baseType).filter { it != baseType }
+    }
+
+    /**
+     * The name of the JSON property that indicates the runtime type of a serialized polymorphic
+     * object.
+     */
+    const val subtypeLabel = "@type"
+
+    /**
+     * Constructs a [RuntimeTypeAdapterFactory] with the given [baseType] and all of its subtypes.
+     */
+    fun <T : Node> polymorphicFactory(baseType: KClass<T>): RuntimeTypeAdapterFactory<T> {
+      val subtypes = findSubtypes(baseType)
+
+      val factory = RuntimeTypeAdapterFactory.of(baseType.java, subtypeLabel)
+      subtypes.forEach { factory.registerSubtype(it.java) }
+
+      return factory
+    }
+
+    /**
+     * Every `abstract`/`sealed` [Node] subclass that is referenced in a field in any other [Node].
+     * Used to register [RuntimeTypeAdapterFactory] instances.
+     */
+    val polymorphicNodeTypes =
+      listOf(
+        Modifier::class,
+        Statement::class,
+        Expression::class,
+        Declaration::class,
+        TypeDeclaration::class,
+        Initializer::class,
+        SwitchStatement.When::class,
+        ElementValue::class
+      )
+  }
+
+  private val gson: Gson
+
+  init {
+    val builder = GsonBuilder().disableHtmlEscaping()
+
+    if (format) {
+      builder.setPrettyPrinting()
+    }
+
+    if (removeLocations) {
+      builder.addSerializationExclusionStrategy(
+        object : ExclusionStrategy {
+          override fun shouldSkipField(field: FieldAttributes): Boolean = false
+          override fun shouldSkipClass(c: Class<*>): Boolean = c == SourceLocation::class.java
+        }
+      )
+    }
+
+    polymorphicNodeTypes.forEach { builder.registerTypeAdapterFactory(polymorphicFactory(it)) }
+
+    gson = builder.create()
+  }
+
+  /** Serializes the given [node] into a JSON string. */
+  fun serialize(node: Node): String {
+    return gson.toJson(node)
+  }
+
+  /** Deserializes the given [JSON string][json] into a node of type [nodeType]. */
```

Comment:
Is this the intended Kdoc formatting?

Reply:
This is what I intended. Could it be improved?

Reply:
I think I misparsed it.  LGTM.